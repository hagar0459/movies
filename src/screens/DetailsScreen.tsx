/**
 * @flow
 * Created by Hagar Abdelghafar on 22.10.2021
 */

import React, { useRef } from 'react';
import
{ StyleSheet, ImageBackground, SafeAreaView, Text, View } from 'react-native';
// import { SharedElement } from 'react-navigation-shared-element';
import * as Animatable from 'react-native-animatable';
import { AppMetrics } from '../../utils/Dimensions';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { styles } from '../../utils/styles';
import { ScrollView } from 'react-native-gesture-handler';
import { IMAGE_BASE_URL } from '../config';

const DetailsScreen = ( { navigation, route } ) =>
{
    const { item, geners } = route.params;
    debugger;
    const buttonRef = useRef();



    return (

        <SafeAreaView>

            <View style={styles.container}>
                {/* <SharedElement id={`item.${ item.id }.image_url`}> */}

                <ImageBackground key={`item.${ item.id }.image_url`}
                    style={{ width: '100%', height: AppMetrics.screenHeight * .4 }} source={{ uri: IMAGE_BASE_URL + item.poster_path }} resizeMode='cover' borderRadius={20} >
                    <Animatable.View
                        key={`item.${ item.id }.view`}
                        ref={buttonRef}
                        animation='fadeIn'
                        duration={600}
                        delay={300}
                        style={[StyleSheet.absoluteFillObject]}
                    >

                        <Icon
                            key={`item.${ item.id }.close`}
                            name='close'
                            size={28}
                            color='#fff'
                            style={{
                                position: 'absolute',
                                top: 40,
                                right: 20,
                                zIndex: 2
                            }}
                            onPress={() =>
                            {
                                buttonRef.current.fadeOut( 100 ).then( () =>
                                {
                                    navigation.goBack();
                                } );
                            }}
                        />
                    </Animatable.View>
                    <View style={styles.languageContainer} key={`item.${ item.id }.langView`}>
                        <Text key={`item.${ item.id }.langText`}
                            style={styles.language}>{item.original_language}</Text>
                    </View>
                </ImageBackground>
                {/* </SharedElement> */}
                <ScrollView showsVerticalScrollIndicator={false}>
                    <View style={{ flexDirection: 'column' }}>
                        {/* <SharedElement id={`item.${ item.id }.cardInfo`}> */}
                        <View key={`item.${ item.id }.infoView`}
                            style={{ width: '100%', justifyContent: 'space-between', flexDirection: 'row', padding: 20 }}>
                            <View style={{ flex: 4 }}>
                                <Text
                                    key={`item.${ item.id }.infoTitle`}
                                    style={styles.title}>{item.title}</Text>
                                <Text key={`item.${ item.id }.infoRealese`}
                                    style={styles.releaseDate}
                                >{item.release_date}</Text>
                            </View>
                            <View style={{ flex: 2, marginHorizontal: 10 }}>
                                <Text key={`item.${ item.id }.infoVotes`}
                                    style={styles.votes}>{item.vote_average}<Text style={styles.votes}>/{item.vote_count}</Text></Text>
                            </View>
                        </View>
                        {/* </SharedElement> */}
                        <Text key={`item.${ item.id }.infoOverview`}
                            style={styles.overview}>{item.overview}</Text>
                        <View style={styles.tagContainer}>
                            {geners.map( ( item, key ) => (
                                <View
                                    key={`item.${ item.id }.tagView`}
                                    style={styles.tagItemContainer}>
                                    <Text key={`item.${ item.id }.tagText`}
                                        style={styles.tagItem}>{item}</Text>
                                </View>
                            ) )}

                        </View>
                    </View>
                </ScrollView>
            </View>
        </SafeAreaView>

    );
};



// DetailsScreen.sharedElements = route =>
// {
//     const { item } = route.params;
//     return [
//         {
//             id: `item.${ item.id }.image_url`,
//             animation: 'move',
//             resize: 'clip'
//         },
//         {
//             id: `item.${ item.id }.cardInfo`,
//             animation: 'fade',
//             resize: 'clip'
//         }
//     ];
// };
export default DetailsScreen;
