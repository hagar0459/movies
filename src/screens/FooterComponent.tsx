/**
 * @flow
 * Created by Hagar Abdelghafar on 22.10.2021
 */


import React from 'react';
import
{
    ActivityIndicator,
    View,
} from 'react-native';
export const FooterComponent = () =>
{

    return (
        <View style={{ height: 20 }}>
            <ActivityIndicator size='large' />
        </View>
    );

};