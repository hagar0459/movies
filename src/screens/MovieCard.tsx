/**
 * @flow
 * Created by Hagar Abdelghafar on 22.10.2021
 */

import React, { FC } from 'react';
import { Platform, Image, StyleSheet, Text, View, TouchableOpacity, TouchableHighlight } from 'react-native';
import type { MovieItemType } from '../redux/models/MovieResultType';
type Props = {
    info: MovieItemType;
    onPress( item: MovieItemType, geners: [] ): void;
    onUpdateFavourite( id: number, state: boolean ): void;
    index: number;
    showFavIcon: boolean
};
import { useDispatch, useSelector } from 'react-redux';
import { SharedElement } from 'react-navigation-shared-element';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { IMAGE_BASE_URL } from '../config';
import { styles } from '../../utils/styles';

export const MovieCard: FC<Props> = ( { info, onPress, onUpdateFavourite, index, showFavIcon }: Props ) =>
{
    debugger;
    const GetFavouriteReducer = useSelector( state => state.GetFavouriteReducer );
    const GetGenerReducer = useSelector( state => state.GetGenerReducer );

    const getGeners = ( geners: any[], reducer: { payload: string | any[]; } ) =>
    {
        let genersNames = [];
        for ( let i = 0; i < geners.length; i++ )
        {
            for ( let d = 0; d < reducer.payload.length; d++ )
            {
                if ( geners[i] === reducer.payload[d].id )
                {
                    genersNames.push( reducer.payload[d].name )
                }
            }
        }
        return genersNames;
    }

    const getFavouriteState = ( id: number ) =>
    {
        if ( GetFavouriteReducer?.payload && GetFavouriteReducer?.payload?.length > 0 )
        {
            for ( let i = 0; i < GetFavouriteReducer?.payload?.length; i++ )
            {
                if ( GetFavouriteReducer?.payload[i]?.id === id )
                {
                    return true
                }
            }
        } else 
        {
            return false
        }
        return false

    }

    const renderContent = () =>
    {


        return (
            <View style={styles.cardContainer} >
                <View>
                    {/* <SharedElement id={`item.${ info.id }.image_url`}> */}

                    <Image source={{ uri: IMAGE_BASE_URL + info.poster_path }}
                        resizeMode="cover" style={styles.cardImage} />
                    {/* </SharedElement> */}
                    {showFavIcon &&
                        <TouchableHighlight
                            underlayColor={Platform.OS === 'ios' ? "transparent" : ''}
                            activeOpacity={0.2}
                            onPress={() => onUpdateFavourite( info.id, getFavouriteState( info.id ) )}>
                            <Icon
                                style={styles.cardFavIcon}
                                name="favorite"
                                color={getFavouriteState( info.id ) ? 'red' : 'grey'}
                                size={30}
                            ></Icon>
                        </TouchableHighlight>
                    }
                    <View style={styles.languageContainer}>
                        <Text style={styles.language}>{info.original_language}</Text>
                    </View>
                </View>
                {/* <SharedElement id={`item.${ info.id }.cardInfo`}> */}
                <View style={{ width: '100%', justifyContent: 'space-between', flexDirection: 'row', marginVertical: 5 }}>
                    <View style={{ flex: 4 }}>
                        <Text style={styles.title}>{info.title}</Text>
                        <Text style={styles.releaseDate}>{info.release_date}</Text>
                    </View>
                    <View style={{ flex: 2, marginHorizontal: 10, alignItems: 'center' }}>
                        <Text style={styles.votes}>{info.vote_average}</Text>

                    </View>
                </View>
                {/* </SharedElement> */}
            </View>
        );

    };
    return (
        <TouchableOpacity
            testID={'CardView'}
            style={styles.card}
            onPress={() =>
            {
                onPress( info, getGeners( info.genre_ids, GetGenerReducer ) );
            }}>
            {renderContent()}
        </TouchableOpacity>
    );
}
