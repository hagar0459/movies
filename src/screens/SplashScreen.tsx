/**
 * @flow
 * Created by Hagar Abdelghafar on 22.10.2021
 */

import React, { useEffect, useState } from 'react';
import { Image, SafeAreaView, View, } from 'react-native';
import { CreatTokenAction } from '../redux/actions/actions'
import { useDispatch } from 'react-redux';
import AppImage from '../../imgs';
import { styles } from '../../utils/styles';


const SplashScreen = ( { navigation } ) =>
{
    const dispatch = useDispatch();

    const [start, setstart] = useState( false )
    if ( start )
    {
        navigation.navigate( 'Tabs' )
    }

    useEffect( () =>
    {
        timeoutHandle = setTimeout( () =>
        {
            dispatch( CreatTokenAction() );
            setstart( true )
        }, 4000 );

    }, [] );
    return (
        <SafeAreaView >

            <View style={[styles.container, {
                justifyContent: 'center',
                alignItems: 'center'
            }]}>
                <Image source={AppImage.splash} resizeMode='contain' style={styles.splashImage} />
            </View>
        </SafeAreaView >
    );
};



export default SplashScreen;