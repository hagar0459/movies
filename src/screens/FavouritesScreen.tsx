/**
 * @flow
 * Created by Hagar Abdelghafar on 22.10.2021
 */

import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { SafeAreaView, FlatList, ListRenderItemInfo, View, BackHandler } from 'react-native';
import { MovieItemType } from '../redux/models/MovieResultType';
import { MovieCard } from './MovieCard';
import { updateFavouriteAction, GetFavouriteList } from '../redux/actions/actions'
import { FooterComponent } from './FooterComponent';
import { styles } from '../../utils/styles';

export const FavouritesScreen = ( { navigation } ) =>
{
    const dispatch = useDispatch();
    const GetFavouriteReducer = useSelector( state => state.GetFavouriteReducer );
    const GetUserDetailsReducer = useSelector( state => state.GetUserDetailsReducer );
    const CreateSessionReducer = useSelector( state => state.CreateSessionReducer );
    const [favourits, setFavourites] = useState( [] )
    const [loading, setLoading] = useState( true );
    const [currentPage, setCurrentPage] = useState( 1 )
    const [sessionId, setSessionId] = useState( null );
    const [accountId, setAccountId] = useState( null );

    const loadMore = () =>
    {

        if ( loading || currentPage === GetFavouriteReducer?.pages )
        {
            return;
        } else if ( GetFavouriteReducer.payload && GetFavouriteReducer.payload.length > 0 )
        {
            dispatch( GetFavouriteList( accountId, currentPage + 1, '' + sessionId ) )
            setCurrentPage( currentPage + 1 );
            setLoading( true )
        }
    };
    useEffect( () =>
    {
        if ( GetUserDetailsReducer.payload && !GetUserDetailsReducer.loading )
        {
            setAccountId( GetUserDetailsReducer.payload )
        }
    }, [GetUserDetailsReducer.loading] );
    useEffect( () =>
    {
        if ( CreateSessionReducer.payload && !CreateSessionReducer.loading )
        {
            setSessionId( CreateSessionReducer.payload )
        }
    }, [CreateSessionReducer.loading] );

    useEffect( () =>
    {
        if ( GetFavouriteReducer.payload.length > 0 && currentPage === 1 )
        {
            setFavourites( GetFavouriteReducer.payload );
            setLoading( false )
        } else if (
            GetFavouriteReducer.payload.length > 0
        )
        {
            setFavourites( [...favourits, ...GetFavouriteReducer.payload] );
            setLoading( false )
        }

    }, [GetFavouriteReducer.payload] );


    const renderFavourite = ( { item, index, }: ListRenderItemInfo<MovieItemType> ) =>
    {
        return (
            <MovieCard
                info={item}
                showFavIcon={true}
                onPress={( item: any, geners: [] ) =>
                {
                    navigation.navigate( 'MovieDetails', {
                        item: item,
                        geners: geners
                    } );
                }}
                onUpdateFavourite={( movie_id, fav_state ) =>
                {
                    if ( accountId && sessionId )
                    {
                        dispatch( updateFavouriteAction( accountId, '' + movie_id, !fav_state, sessionId ) )
                    }
                }}
                index={index}
            />
        );
    };
    useEffect( () =>
    {
        const backHandler = BackHandler.addEventListener( 'hardwareBackPress', () => true )
        return () => backHandler.remove()
    }, [] )
    return (
        <SafeAreaView style={{ flex: 1 }} >

            <View style={{ flex: 1 }}>
                <FlatList
                    keyboardShouldPersistTaps='never'
                    style={styles.moviesList}
                    contentContainerStyle={styles.moviesListContainer}
                    data={favourits}
                    keyExtractor={item => `${ item.id }`}
                    renderItem={renderFavourite}
                    onEndReached={loadMore}
                    ListFooterComponent={FooterComponent}
                />
            </View>
        </SafeAreaView>
    );
};



