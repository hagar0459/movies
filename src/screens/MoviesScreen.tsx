/**
 * @flow
 * Created by Hagar Abdelghafar on 22.10.2021
 */

import React, { FC, useState, useRef, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FlatList, StyleSheet, SafeAreaView, TextInput, View, ListRenderItemInfo, BackHandler } from 'react-native';
import { MovieItemType } from '../redux/models/MovieResultType';
import { MovieCard } from './MovieCard';
import { GetMoviesListAction, updateFavouriteAction, GetSearchAction } from '../redux/actions/actions'
import { FooterComponent } from './FooterComponent';
import { styles } from '../../utils/styles';

export const MoviesScreen: FC = ( { navigation } ) =>
{
    const dispatch = useDispatch();

    let ref = useRef<FlatList<MovieItemType>>( null );
    const GetMoviesReducer = useSelector( state => state.GetMoviesReducer );
    const GetUserDetailsReducer = useSelector( state => state.GetUserDetailsReducer );
    const CreateSessionReducer = useSelector( state => state.CreateSessionReducer );

    const [searchTxt, setSearchTxt] = useState( '' );
    const [movies, setMovies] = useState( [] )
    const [loading, setLoading] = useState( true );
    const [currentPage, setCurrentPage] = useState( 1 )
    const [sessionId, setSessionId] = useState( null );
    const [accountId, setAccountId] = useState( null );

    //load more for movies
    const loadMore = () =>
    {
        if ( loading || currentPage === GetMoviesReducer?.pages )
        {
            return;
        } else if ( GetMoviesReducer?.payload && GetMoviesReducer?.payload?.length > 0 )
        {
            dispatch( GetMoviesListAction( currentPage + 1, '' + sessionId ) )
            setCurrentPage( currentPage + 1 );
            setLoading( true )
        }
    };

    useEffect( () =>
    {
        if ( GetUserDetailsReducer?.payload && !GetUserDetailsReducer?.loading )
        {
            setAccountId( GetUserDetailsReducer?.payload )
        }
    }, [GetUserDetailsReducer?.loading] );
    useEffect( () =>
    {
        const backHandler = BackHandler.addEventListener( 'hardwareBackPress', () => true )
        return () => backHandler.remove()
    }, [] )
    useEffect( () =>
    {
        if ( CreateSessionReducer?.payload && !CreateSessionReducer?.loading )
        {
            setSessionId( CreateSessionReducer?.payload )
        }
    }, [CreateSessionReducer?.loading] );
    useEffect( () =>
    {
        if ( GetMoviesReducer?.payload?.length > 0 && currentPage === 1 )
        {
            setMovies( GetMoviesReducer?.payload );
            setLoading( false )
        } else if ( GetMoviesReducer?.payload?.length > 0 )
        {
            setMovies( [...movies, ...GetMoviesReducer?.payload] );
            setLoading( false )
        }
    }, [GetMoviesReducer?.payload] );

    const renderCharacter = ( { item, index, }: ListRenderItemInfo<MovieItemType> ) =>
    {
        return (
            <MovieCard
                info={item}
                showFavIcon={true}
                onPress={( item: any, geners: [] ) =>
                {
                    navigation.navigate( 'MovieDetails', {
                        item: item,
                        geners: geners
                    } );
                }}
                onUpdateFavourite={( movie_id, fav_state ) =>
                {
                    if ( accountId && sessionId )
                    {
                        dispatch( updateFavouriteAction( accountId, '' + movie_id, !fav_state, sessionId ) )
                    }
                }}
                index={index}
            />
        );
    };
    return (
        <SafeAreaView>
            <View style={styles.inputContainer}>
                <TextInput
                    style={styles.input}
                    placeholder='search...'
                    onChangeText={( text ) =>
                    {
                        ref.current?.scrollToOffset( { animated: true, offset: 0 } );
                        setSearchTxt( text );
                        setCurrentPage( 1 )
                        if ( text && text.length > 0 )
                        {
                            dispatch( GetSearchAction( text, 1 ) )
                        } else
                        {
                            dispatch( GetMoviesListAction( 1, '' + sessionId ) )

                        }
                    }}
                    autoCorrect={false}
                    value={searchTxt}
                />
            </View>
            <FlatList
                keyboardShouldPersistTaps='never'
                ref={ref}
                style={styles.moviesList}
                contentContainerStyle={styles.moviesListContainer}
                data={movies}
                keyExtractor={item => `${ item.id }`}
                renderItem={renderCharacter}
                onEndReached={loadMore}
                ListFooterComponent={FooterComponent}
            />
        </SafeAreaView>
    );
};
