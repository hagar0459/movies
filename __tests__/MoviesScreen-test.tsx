/**
 * @format
 */

import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { MoviesScreen } from '../src/screens/MoviesScreen';

// Note: test renderer must be required after react-native.
describe( '<MovieCard /> tests', () =>
{
    const initialState = { output: 10 }
    const mockStore = configureStore()
    let store, wrapper

    it( 'Render the component correctly ', () =>
    {
        store = mockStore( initialState )


        const navigate = jest.fn();
        const tree = renderer.create( <Provider store={store}><MoviesScreen navigation={{ navigate }} /></Provider> ).toJSON();
        expect( tree ).toMatchSnapshot();
    } );
} );