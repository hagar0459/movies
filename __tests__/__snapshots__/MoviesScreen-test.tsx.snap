// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> tests Render the component correctly  1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "height": 60,
        "width": "100%",
      }
    }
  >
    <TextInput
      autoCorrect={false}
      onChangeText={[Function]}
      placeholder="search..."
      style={
        Object {
          "backgroundColor": "#dcdcdc",
          "borderRadius": 10,
          "flex": 1,
          "fontSize": 18,
          "marginBottom": 10,
          "marginHorizontal": "5%",
          "marginTop": 10,
          "paddingHorizontal": 10,
          "width": "90%",
        }
      }
      value=""
    />
  </View>
  <RCTScrollView
    ListFooterComponent={[Function]}
    contentContainerStyle={
      Object {
        "backgroundColor": "#fafafa",
        "marginHorizontal": 8,
        "paddingVertical": 8,
      }
    }
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="never"
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#fafafa",
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "height": 20,
            }
          }
        >
          <ActivityIndicator
            size="large"
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
